File: src/parser/expression.cc
#include "monkey/parser/expression.hh"

#include <cstdlib>
#include <memory>
#include <string_view>
#include <utility>

#include <cassert>

#include "monkey/common.hh"
#include "monkey/lexer/token.hh"
#include "monkey/parser/parser.hh"

BEGIN_MONKEY_NAMESPACE

std::unique_ptr<Expression>
parse_grouped_expression(Parser& parser) noexcept
{
    parser.next();
    std::unique_ptr<Expression> e = parser.parse_expression(Precedence::Lowest);

    if (!parser.next_token_is(TokenType::RParenthesis)) {
        return nullptr;
    }

    return e;
}

// ----------------------------------------------------------------------------

IdentifierExpression::IdentifierExpression(const std::string_view _value) noexcept
  : value{ std::move(_value) }
{
}

std::unique_ptr<IdentifierExpression>
IdentifierExpression::parse(Parser& parser) noexcept
{
    assert(parser.current_token_is(TokenType::Identifier));
    std::string_view value{ parser.current_token().value().literal };
    return std::make_unique<IdentifierExpression>(std::move(value));
}

// ----------------------------------------------------------------------------

IntegerLiteral::IntegerLiteral(int64_t _value) noexcept
  : value{ _value }
{
}

std::unique_ptr<IntegerLiteral>
IntegerLiteral::parse(Parser& parser) noexcept
{
    assert(parser.current_token_is(TokenType::Integer));
    int64_t value = atoll(parser.current_token().value().literal.data());
    return std::make_unique<IntegerLiteral>(value);
}

// ----------------------------------------------------------------------------

PrefixExpression::PrefixExpression(const TokenType _type,
                                   const std::string_view _op,
                                   std::unique_ptr<Expression> _right) noexcept
  : type{ _type }
  , op{ std::move(_op) }
  , right{ std::move(_right) }
{
}

std::unique_ptr<PrefixExpression>
PrefixExpression::parse(Parser& parser) noexcept
{
    assert(parser.current_token().has_value());
    const Token& c = parser.current_token().value();
    auto expression = std::make_unique<PrefixExpression>(c.type, c.literal, nullptr);
    parser.next();
    expression->right = parser.parse_expression(Precedence::Prefix);
    return expression;
}

// ----------------------------------------------------------------------------

InfixExpression::InfixExpression(const TokenType _type,
                                 std::unique_ptr<Expression> _left,
                                 const std::string_view _op,
                                 std::unique_ptr<Expression> _right) noexcept
  : type{ _type }
  , left{ std::move(_left) }
  , op{ std::move(_op) }
  , right{ std::move(_right) }
{
}

std::unique_ptr<InfixExpression>
InfixExpression::parse(Parser& parser, std::unique_ptr<Expression> _left) noexcept
{
    assert(parser.current_token().has_value());
    const Token& current = parser.current_token().value();
    Precedence precedence = Parser::get_precedence(current.type);
    parser.next();
    std::unique_ptr<Expression> right = parser.parse_expression(precedence);
    return std::make_unique<InfixExpression>(current.type, std::move(_left), current.literal,
                                             std::move(right));
}

// ----------------------------------------------------------------------------

std::unique_ptr<BlockStatement>
BlockStatement::parse(Parser& parser) noexcept
{
    assert(parser.current_token().has_value());
    auto block = std::make_unique<BlockStatement>();

    while (parser.current_token().has_value() && !parser.current_token_is(TokenType::RBrace)) {
        std::unique_ptr<Statement> statement = parser.parse_statement();

        if (statement.get() != nullptr) {
            block->statements.push_back(std::move(statement));
        }

        parser.next();
    }

    return block;
}

// ----------------------------------------------------------------------------

IfExpression::IfExpression(std::unique_ptr<Expression> _condition,
                           std::unique_ptr<BlockStatement> _consequence,
                           std::unique_ptr<BlockStatement> _alternative) noexcept
  : condition{ std::move(_condition) }
  , consequence{ std::move(_consequence) }
  , alternative{ std::move(_alternative) }
{
}

std::unique_ptr<IfExpression>
IfExpression::parse(Parser& parser) noexcept
{
    assert(parser.current_token_is(TokenType::If));

    if (!parser.next_token_is(TokenType::LParenthesis)) {
        return nullptr;
    }

    parser.next();
    std::unique_ptr<Expression> condition = parser.parse_expression(Precedence::Lowest);

    if (!parser.next_token_is(TokenType::RParenthesis)) {
        return nullptr;
    }

    parser.next();

    if (!parser.next_token_is(TokenType::LBrace)) {
        return nullptr;
    }

    parser.next();
    std::unique_ptr<BlockStatement> consequence = BlockStatement::parse(parser);
    std::unique_ptr<BlockStatement> alternative = nullptr;

    if (parser.next_token_is(TokenType::Else)) {
        parser.next();

        if (!parser.next_token_is(TokenType::LBrace)) {
            return nullptr;
        }

        parser.next();
        alternative = BlockStatement::parse(parser);
    }

    return std::make_unique<IfExpression>(std::move(condition), std::move(consequence),
                                          std::move(alternative));
}

// ----------------------------------------------------------------------------

FunctionLiteral::FunctionLiteral(std::unique_ptr<BlockStatement> _body) noexcept
  : body{ std::move(_body) }
{
}

std::unique_ptr<FunctionLiteral>
FunctionLiteral::parse(Parser& parser) noexcept
{
    assert(parser.current_token_is(TokenType::Function));

    if (!parser.next_token_is(TokenType::LParenthesis)) {
        return nullptr;
    }

    parser.next();
    auto fn = std::make_unique<FunctionLiteral>(/*_body=*/nullptr);
    parse_function_parameters(parser, *fn);

    if (!parser.next_token_is(TokenType::LBrace)) {
        return nullptr;
    }

    parser.next();
    fn->body = BlockStatement::parse(parser);

    return fn;
}

void
parse_function_parameters(Parser& parser, FunctionLiteral& fn) noexcept
{
    assert(parser.current_token_is(TokenType::LParenthesis));

    // Handle case where there are no function parameters.
    if (parser.next_token_is(TokenType::RParenthesis)) {
        parser.next();
        return;
    }

    parser.next();
    const std::optional<Token>& current = parser.current_token();
    std::vector<std::string_view> parameter_names{};

    while (parser.next_token_is(TokenType::Comma)) {
        parameter_names.emplace_back(current.value().literal);
        parser.next(); // Identifier => Comma
        parser.next(); // Comma => Next Identifier
    }

    if (!parser.next_token_is(TokenType::RParenthesis)) {
        return;
    }

    for (auto& name : parameter_names) {
        fn.parameters.emplace_back(std::make_unique<IdentifierExpression>(std::move(name)));
    }

    // TODO: There is currently no way to tell whether the function returned
    // because there were no function parameters, or if it left because there
    // was an error with the parameter names...
}

END_MONKEY_NAMESPACE

File: src/ast/expression.rs
use crate::token::{Token, TokenKind};

use super::parser::{Parse, Parser, ParserError};
use super::statement::Block;
use super::Node;

#[derive(Debug)]
pub enum Expression {
    Identifier(Identifier),
    IntegerLiteral(IntegerLiteral),
    Prefix(Prefix),
    Infix(Infix),
    Boolean(Boolean),
    If(If),
    FunctionLiteral(FunctionLiteral),
    Call(Call),
}

impl From<Node> for Expression {
    fn from(value: Node) -> Self {
        match value {
            Node::Expression(expression) => expression,
            _ => panic!("expected node to be an expression"),
        }
    }
}

// ┌────────────────────────────────┐
// │ Implementations for Identifier │
// └────────────────────────────────┘

#[derive(Debug)]
pub struct Identifier {
    pub token: Token,
}

impl From<Identifier> for Expression {
    fn from(value: Identifier) -> Self {
        Expression::Identifier(value)
    }
}

impl From<Identifier> for Node {
    fn from(value: Identifier) -> Self {
        Expression::Identifier(value).into()
    }
}

impl Parse<'_> for Identifier {
    fn parse(parser: &mut Parser<'_>) -> Result<Node, ParserError> {
        let node = parser
            .expect_token(TokenKind::Identifier)
            .map(|token| Self { token })?
            .into();

        Ok(node)
    }
}

// ┌─────────────────────────────────────┐
// │ Implementations for Integer Literal │
// └─────────────────────────────────────┘

#[derive(Debug)]
pub struct IntegerLiteral {
    pub token: Token,
    pub value: i64,
}

impl From<IntegerLiteral> for Expression {
    fn from(value: IntegerLiteral) -> Self {
        Expression::IntegerLiteral(value)
    }
}

impl From<IntegerLiteral> for Node {
    fn from(value: IntegerLiteral) -> Self {
        Expression::IntegerLiteral(value).into()
    }
}

impl Parse<'_> for IntegerLiteral {
    fn parse(parser: &mut Parser<'_>) -> Result<Node, ParserError> {
        let node = parser
            .expect_token(TokenKind::Integer)
            .map(|token| {
                let value = token.literal.clone().parse::<i64>().unwrap();
                Self { token, value }
            })?
            .into();

        Ok(node)
    }
}

// ┌────────────────────────────┐
// │ Implementations for Prefix │
// └────────────────────────────┘

#[derive(Debug)]
pub struct Prefix {
    pub token: Token,
    pub right: Box<Expression>,
}

impl From<Prefix> for Expression {
    fn from(value: Prefix) -> Self {
        Expression::Prefix(value)
    }
}
impl From<Prefix> for Node {
    fn from(value: Prefix) -> Self {
        Expression::Prefix(value).into()
    }
}

impl Parse<'_> for Prefix {
    fn parse(_parser: &mut Parser<'_>) -> Result<Node, ParserError> {
        todo!()
    }
}

// ┌───────────────────────────┐
// │ Implementations for Infix │
// └───────────────────────────┘

#[derive(Debug)]
pub struct Infix {
    pub token: Token,
    pub left: Box<Expression>,
    pub right: Box<Expression>,
}

impl From<Infix> for Expression {
    fn from(value: Infix) -> Self {
        Expression::Infix(value)
    }
}

impl From<Infix> for Node {
    fn from(value: Infix) -> Self {
        Expression::Infix(value).into()
    }
}

impl Parse<'_> for Infix {
    fn parse(_parser: &mut Parser<'_>) -> Result<Node, ParserError> {
        todo!()
    }
}

// ┌─────────────────────────────┐
// │ Implementations for Boolean │
// └─────────────────────────────┘

#[derive(Debug)]
pub struct Boolean {
    pub token: Token,
    pub value: bool,
}

impl From<Boolean> for Expression {
    fn from(value: Boolean) -> Self {
        Expression::Boolean(value)
    }
}

impl From<Boolean> for Node {
    fn from(value: Boolean) -> Self {
        Expression::Boolean(value).into()
    }
}

impl Parse<'_> for Boolean {
    fn parse(parser: &mut Parser<'_>) -> Result<Node, ParserError> {
        let token = parser.token.clone();
        let value = token.kind == TokenKind::True;
        let node = Self { token, value }.into();
        Ok(node)
    }
}

// ┌────────────────────────┐
// │ Implementations for If │
// └────────────────────────┘

#[derive(Debug)]
pub struct If {
    pub token: Token,
    pub condition: Box<Expression>,
    pub statement: Box<Expression>,
    pub alternative: Option<Box<Expression>>,
}

impl From<If> for Expression {
    fn from(value: If) -> Self {
        Expression::If(value)
    }
}

impl From<If> for Node {
    fn from(value: If) -> Self {
        Expression::If(value).into()
    }
}

impl Parse<'_> for If {
    fn parse(_parser: &mut Parser<'_>) -> Result<Node, ParserError> {
        todo!()
    }
}

// ┌──────────────────────────────────────┐
// │ Implementations for Function Literal │
// └──────────────────────────────────────┘

#[derive(Debug)]
pub struct FunctionLiteral {
    pub token: Token,
    pub parameters: Vec<Identifier>,
    pub body: Block,
}

impl From<FunctionLiteral> for Expression {
    fn from(value: FunctionLiteral) -> Self {
        Expression::FunctionLiteral(value)
    }
}

impl From<FunctionLiteral> for Node {
    fn from(value: FunctionLiteral) -> Self {
        Expression::FunctionLiteral(value).into()
    }
}

impl Parse<'_> for FunctionLiteral {
    fn parse(_parser: &mut Parser<'_>) -> Result<Node, ParserError> {
        todo!()
    }
}

// ┌──────────────────────────┐
// │ Implementations for Call │
// └──────────────────────────┘

#[derive(Debug)]
pub struct Call {
    pub token: Token,
    pub function: Box<Expression>,
    pub arguments: Vec<Box<Expression>>,
}

impl From<Call> for Expression {
    fn from(value: Call) -> Self {
        Expression::Call(value)
    }
}

impl From<Call> for Node {
    fn from(value: Call) -> Self {
        Expression::Call(value).into()
    }
}

impl Parse<'_> for Call {
    fn parse(_parser: &mut Parser<'_>) -> Result<Node, ParserError> {
        todo!()
    }
}
